[V3 Streams] Fix for #1132

VicenteRD opened the pull request:
### Type

Bugfix

### Description of the changes

Fixes #1132 : `self.streams` being a `Task` object instead of a list. 
Also replicated for `self.communities`, since the issue is the same 
(although untested, it can be inferred).

VicenteRD committed:
Fix for #1132

Fixes `self.streams` being a `Task` object instead of a list. 
Also replicated for `self.communities`, since the issue is the same 
(although untested, it can be inferred).

palmtree5 commented:
Just checked, definitely does happen with communities too, so good call on fixing it for that too

VicenteRD committed:
Merge branch 'V3/develop' into patch-1

Tobotimus commented on redbot/cogs/streams/main.py:
```diff
@@ -41,8 +41,9 @@ def __init__(self, bot: Red):
 
-        self.streams = self.bot.loop.create_task(self.load_streams())
-        self.communities = self.bot.loop.create_task(self.load_communities())
+        self.bot.loop.create_task(self.load_streams())
```
`self.streams` and `self.communities` must be initialised as an iterable before `_stream_alerts()` is called, otherwise it will raise an `AttributeError` or `type is not iterable` error. This is possible in the current scenario because these three tasks are scheduled to run asynchronously, and not linearly.

You could do this just by initialising them as empty lists before scheduling their `load_X` functions, or otherwise.

VicenteRD replied:
I thought about an empty pre-initialization too, but I didn't think it'd be needed...

Now that I look at `_stream_alerts`, you might want the 2 previous tasks to complete before that happens though (So the first run of that method is done with a complete list of streams and communities).
So the 2 possible solutions are empty pre-init, or waiting for the tasks. You tell me :)

Tobotimus replied:
~~I don't think waiting for tasks is a good idea here, as it will have blocking effects.~~ You could create another async function which awaits the first two tasks before creating the `self._stream_alerts()` task, so you can guarantee linearity. Then you only have to use `create_task` once in the `__init__`.

Edit: I think I misunderstood what you meant by "waiting for the tasks".

Tobotimus reviewed (changes requested):
Other than this one change, everything looks good to me.

VicenteRD committed:
Merge branch 'V3/develop' into patch-1

VicenteRD committed:
Merge branch 'V3/develop' into patch-1

VicenteRD committed:
Implement sub-method to avoid init errors

The new method waits for `streams` and `communities` to be initialized
before running the `_stream_alerts` method, avoiding possible errors
due to lists being uninitialized

VicenteRD committed:
Merge branch 'V3/develop' into patch-1

Tobotimus commented on redbot/cogs/streams/main.py:
```diff
@@ -41,9 +41,12 @@ def __init__(self, bot: Red):
+
+    async def _initialize_lists(self):
+        await self.bot.loop.create_task(self.load_streams())
+        await self.bot.loop.create_task(self.load_communities())
```
This will fail since `create_task` isn't a coroutine function. You only need to await the coroutines themselves, i.e. `await self.load_streams()` and `await self.load_communities()`.

VicenteRD replied:
How am I so stupid >.< Sorry, I'll fix tomorrow morning

VicenteRD committed:
Fix streams and communities loading

VicenteRD committed:
Merge remote-tracking branch 'origin/patch-1' into patch-1

VicenteRD committed:
Delete IDE config file

VicenteRD committed:
Merge branch 'V3/develop' into patch-1

Tobotimus reviewed (approved):
Tested and it works :+1:

VicenteRD committed:
Merge branch 'V3/develop' into patch-1

tekulvw committed:
Merge branch 'V3/develop' into patch-1

tekulvw commented:
Please DM me on Discord to receive your "Contributor" role!
